src/controller.cpp

```diff
@@ -0,0 +1,1230 @@
+  turning_radius_inv_vector_.clear();
+  turning_radius_inv_vector_.resize(global_plan_tf_.size());
+  turning_radius_inv_vector_[global_plan_tf_.size() - 1] = 0.0;
+  tf2::Transform deltaPlan;
+  // We define segment0 to be the segment connecting pose0 and pose1.
+  // Hence, when picking the starting path's pose, we mean to start at the segment connecting that and the next pose.
+  for (int idx_path = global_plan_tf_.size() - 2; idx_path >= 0; --idx_path)
+  {
+    /* Get distance to segment to determine if this is the segment to start at */
+    dist_to_segment = distToSegmentSquared(current_tf2, global_plan_tf_[idx_path], global_plan_tf_[idx_path + 1]);
+    // Calculate 3D distance, since current_tf2 might have significant z-offset and roll/pitch values w.r.t. path-pose
+    // When not doing this, we're brutely projecting in robot's frame and might snap to another segment!
+    if (dist_to_segment < minimum_distance_to_path)
+    {
+      minimum_distance_to_path = dist_to_segment;
+      controller_state_.current_global_plan_index = idx_path;
+    }
+
+    /* Create distance and turning radius vectors once for usage later */
+    deltaPlan = global_plan_tf_[idx_path].inverseTimes(global_plan_tf_[idx_path + 1]);
+    double dpX = deltaPlan.getOrigin().x();
+    double dpY = deltaPlan.getOrigin().y();
+    iterative_dist_to_goal += hypot(dpX, dpY);
+    distance_to_goal_vector_[idx_path] = iterative_dist_to_goal;
+    // compute turning radius based on trigonometric analysis
+    // radius such that next pose is connected from current pose with a semi-circle
+    double dpXY2 = dpY*dpY + dpX*dpX;
+    if (dpXY2 < FLT_EPSILON)
+    {
+      turning_radius_inv_vector_[idx_path] = std::numeric_limits<double>::infinity();
+    }
+    else
+    {
+      //  0.5*dpY*( 1 + dpX*dpX/(dpY*dPY) );
+      // turning_radius_vector[idx_path] = 0.5*(1/dpY)*( dpY*dpY + dpX*dpX );
+      turning_radius_inv_vector_[idx_path] = 2*dpY/dpXY2;
+    }
+    ROS_DEBUG("turning_radius_inv_vector[%d] = %f", idx_path, turning_radius_inv_vector_[idx_path]);
+  }
+
+  // Set initial velocity
+  switch (local_config_.init_vel_method)
+  {
+    case Pid_Zero:
+      reset();
+      break;
+    case Pid_Odom:
+      reset();
+      controller_state_.current_x_vel = odom_twist.linear.x;
+      controller_state_.current_yaw_vel = odom_twist.angular.z;
+      ROS_INFO("Resuming on odom velocity x: %f, yaw: %f", odom_twist.linear.x, odom_twist.angular.z);
+      break;
+    default:
+      ROS_DEBUG("Internal controller_state stays valid");
+      break;
+  }
+
+  // When velocity error is too big reset current_x_vel
+  if (fabs(odom_twist.linear.x - controller_state_.current_x_vel) > max_error_x_vel_)
+  {
+    // TODO(clopez/mcfurry/nobleo): Give feedback to higher level software here
+    ROS_WARN("Large control error. Current_x_vel %f / odometry %f",
+            controller_state_.current_x_vel, odom_twist.linear.x);
+  }
```

Comment:
"`reset current_x_vel`"? I don't see this happening...

Reply:
There was a reset at some point, but then it introduced a undesired behavior in one of the applications, we should remove or rewrite the comment.

Reply:
We can get rid of this condition and warning, because it is doesn't do anything functionally, and a similar check is performed in the `update` method here:
https://github.com/nobleo/path_tracking_pid/blob/61c6cc828bd98dddd8cc14cbabdd62353a89ad3c/src/controller.cpp#L642-L649